### ISSUE TITLE
Piwik block by AdBlock even after renmaing piwik.php piwik.js

### ISSUE THREAD

**Issue by sambuev** [piwik/piwik#7364]
Hello, I have Piwik installed on my hosting with secret address: interes.site.com/sm/
and I have renamed piwik.php piwik.js to sm.php sm.js and I though it cannot by caught by AdBlock but unfortunately AdBlock on my Safari browser easily blocking Piwik and many other users with their Adblock (on Chrome, Firefox) out of my view :( 

![screen shot 2015-03-04 at 18 21 38](https://cloud.githubusercontent.com/assets/6326685/6495985/b28d4c06-c29c-11e4-924b-53789685e875.png)

What can I do to avoid blocking Piwik? 
Is it possible to avoid it?

**Comment by MESWEB** [comment]
I got this same error but I think this is difficulty for dev piwik

**Comment by kylekatarnls** [comment]
I use rewriting and replace /piwik.php with juste /p. Works like a charm.

**Comment by MESWEB** [comment]
Can You give any solution? I trying this but without success - https://www.codelibrary.me/tag/ad-block/

**Comment by kylekatarnls** [comment]
This is what I did (but with a proxy-pass in the same domain).

First remove "analytics" in the URL when you enter the rewrite rule and any word that can be seen as ads.

Then if it's not enough, use a proxypass (http://httpd.apache.org/docs/2.2/fr/mod/mod_proxy.html) on your own domain, so your URL to call Piwik will just look like "/p" and "/j"

```
(function() {
    _paq.push(['setTrackerUrl', '/p']);
    _paq.push(['setSiteId', 1]);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src='/j'; s.parentNode.insertBefore(g,s);
  })();
```

**Comment by Joey3000** [comment]
Guys, have you tried just replacing "piwik.php" and "piwik.js" with "js/", as described in https://github.com/piwik/piwik/blob/master/js/README.md? No .htaccess rewriting, file renaming, etc. needed.

See also:
http://forum.piwik.org/read.php?2,83940,125929
http://forum.piwik.org/read.php?4,14971

I've been using that for years. Just make sure that the "piwik" string doesn't appear anywhere in the URL (i.e. domain name, directories). After coming across this issue report, I've just tried it again with ABE on Firefox and it doesn't block it. I checked ABE lists again, and it only checks for the "piwik" string anywhere in the URL.

**Comment by MESWEB** [comment]
Non of all solution what You give not working with uBlock! I think the uBlock is hard to defeat. Only solution is use any server-side tracker. Piwik should use server-side tracking by default.

**Comment by Joey3000** [comment]
@MESWEB From https://addons.mozilla.org/en-us/firefox/addon/ublock-origin/:

```
Out of the box, these lists of filters are loaded and enforced:

- EasyList
- Peter Lowe’s Ad server list
- EasyPrivacy
- Malware domains
```

Any chance your Piwik URL doesn't pass any of those?

**Comment by Joey3000** [comment]
There is following investigation necessary:

* If the lists can include not just Piwik installation URLs, but also URL parameters. (I tried the lists provided in the original issue post with ABP on Firefox and Piwik was not blocked. But I haven't checked if the "Peter Lowe’s Ad server list" also used by ublock includes URL parameters. Or if any other list could do that.)
* If the ability by the plugin to check the parameters is browser dependent (e.g. due to some browser restrictions). Since the original issue post refers to Safari.

And if the block can be URL parameter based, it could be the time for a Piwik plugin where the user could freely change them. (I don't know how deep into the rabbit hole one would need to go to do that.)

**Comment by kylekatarnls** [comment]
@MESWEB Piwik was using server-side (it was called phpMyVisites), but for many reasons (performances, security) it has been switched to client-side and it's a good thing. I use proxypass + url-rewriting (it's exactly equivalent to the solution mentionned by @Joey3000, but it is faster because it does not use PHP) and uBlock does not block this solution.

**Comment by MESWEB** [comment]
Yes I know it and I try changing url or filenames and dynamic changed filenames with .htaccess with RewriteRuele but without effect. I saw something strange - Any of my browser (Opera, Chrome, Waterfox) hasn't be counted by piwik but when Author of the PIWIK goes on my site I see his log in PIWIK. So my question is WTF? I'm not only one with this strange bug.

**Comment by kylekatarnls** [comment]
Maybe you missclicked on uBlock and add a filter in your browser try to reset uBlock. Else, if all your stuff is on the same domain, it should not be blocked as uBlock does not search in the response body but only on request URLs.

**Comment by bdore** [comment]
Ok so I was able to go around Opera's built-in adblocker with some major changes to PiWik. I'm still testing to see if it holds together after deploy. This was tested only on a local server and PiWik self-hosted:
- Opera has an option to send "do no track" requests which are respected by PiWik. I did not find a workaround for that but I figure if the user goes that far I might as well respect that.

Changes I made to the PiWik installation files:

- rename installation directory to *p*
- *p/piwik.js* renamed to *p.js*
- *p/piwik.php* renamed to *p.php*
- main script array *_paq* renamed to *yourArray*. Not sure if this is necessary, try without this first
	- find *_paq*, replace with *yourArray* in *p.js*
- find all occurrences of *action_name* in files under *p/* (*action_name* is a request parameter, this was being picked up by the blocker)
	- replace ocurrences with *yourParameter* in:
		- *p.js*
		- *core/Tracker/ActionPageview.php*
                - *core/Tracker/Request.php*
		- *js/piwik.js*
		- *js/piwik.min.js*
		- *vendor/piwik/piwik-php-tracker/PiwikTracker.php*

Do all this at your own risk, this is totally experimental. This probably breaks something in PiWik. Not thoroughly tested.

**Comment by novakin** [comment]
@bdore would be great to be able to do that trough a plugin

**Comment by collinbarrett** [comment]
It used to be that using the [proxy](https://github.com/piwik/tracker-proxy) and renaming piwik.php would work, but [this recent commit](https://github.com/easylist/easylist/commit/3a7b9efe481aa3cb1d83ac0fe543c1e04f4d2cb8) to EasyPrivacy makes it trickier. I think the steps that @bdore suggested regarding renaming some of the query parameters would now be required. I don't plan to go down this path anymore, but if anyone wants to use Piwik and really wants to circumvent EasyPrivacy users, that would be the trick. A plugin would certainly be required. Otherwise, the customizations to Piwik could be lost on upgrade.

**Comment by sboesch** [comment]
Keep in mind, instead of renaming any files, you should also be able to create symlinks:

`$ ln -s piwik.js p.js && ln -s piwik.php p.php`

This is much cleaner and at least uBlock seems to allow these requests as far as I checked for now.

**Comment by FabriceSalvaire** [comment]
**A lot of things are wrong here !**

EasyList contains these basic rules :

* /piwik-$domain=~github.com|~piwik.org
* /piwik.$script,domain=~piwik.org
* /piwik.php /piwik/*$domain=~github.com|~piwik.org
* /piwik1.
* /piwik2.js
* /piwik_
* /piwikapi.js
* /piwikC_
* /piwikTracker.
* .php?action_name=
* .php?logRefer=
* .php?logType=
* .php?p=stats&
* .php?ping=
* .php?refcode=
* .php?tracking=

Thus we have just to change in the tracking script

* **piwik.js** to **foo.js**
* **piwik.php** to **foo**

And a two URL rules in the webserver to map them to the original paths.

It will work until the parameter list doesn't contain evident keywords like stat, log, track ... Blockers can try to be clever, but they will never broke the www.

**Comment by arnowelzel** [comment]
If you use the "proxy" script to hide the real URL of your Piwik server (see https://github.com/piwik/tracker-proxy) you can also use that to hide parameter names as well.

When sending the JavaScript, just do a str_replace for "action_name=" to something else, e.g. "aname=" and then rename the $_GET parameter back to the original name. You should also rename this proxy script from "piwik.php" to something else.

So - first replace "action_name" in the provided JavaScript from the Piwik server:

```
        if ($piwikJs = $content) {
            echo str_replace('"action_name="', '"aname="', $piwikJs); 
        } else {
```
And this to accept "aname" as parameter but forward it as "action_name" in the server side CURL request to the Piwik server:

```
if(isset($_GET['aname']) {
        $_GET['action_name'] = $_GET['aname'];
        unset($_GET['aname']);
}
foreach ($_GET as $key => $value) {
    $url .= urlencode($key ). '=' . urlencode($value) . '&';
}
```

But be aware, that even the new parameter name "aname" may once get into filter lists as well - so you may have to use something else.

**Comment by omarr1000** [comment]
@arnoweizel: where do place this code? In which file?
I tried in piwik.php (from tracker-proxy), but no success, nothing is replaced, still have "action_name" 
thx for an advice in advance

**Comment by arnowelzel** [comment]
@omarr1000: then you did something wrong. See the script here, which I use on my own server (https://arnowelzel.de):

https://arnowelzel.de/samples/piwik-tracker-proxy.txt

**Comment by omarr1000** [comment]
thx! now I got it, did write the code at the wrong line )-;

**Comment by londonuk371** [comment]
I had to do 2 things:
 -First the symlinks `$ ln -s piwik.js p.js && ln -s piwik.php p.php` like proposed by @sboesch 
 - Then modifiy piwik.js / piwik.php: 
**piwik:js**
Rename "action_name=" to "_action_name"
**piwik.php**
Added in the first line of piwik.php:
`if(isset($_GET['_action_name'])) { $_GET['action_name'] = $_GET['_action_name']; }`

ublocOrigin was blocking `?action_name=`

**Comment by hparadiz** [comment]
Matomo tracking information should be obfuscated both to avoid being picked up by ad blockers but also for the security of the person being tracked! The current tracking URL is obnoxious with it's GET variables being sent over HTTP in plaintext.

Here's an example:
`https://domain/a.php?action_name=site&idsite=1&rec=1&r=058405&h=11&m=1&s=31&url=https%3A%2F%2Fdomain%2F&_id=9b550a96cfa8d490&_idts=1523890834&_idvc=1&_idn=0&_refts=0&_viewts=1523890834&send_image=1&pdf=1&qt=0&realp=0&wma=0&dir=0&fla=0&java=0&gears=0&ag=0&cookie=1&res=1920x1200&gt_ms=27&pv_id=Zf3r32`

Matomo _should_ generate a short random string and automatically create this `.htaccess` file if mod_rewrite is detected:
```
RewriteEngine On

RewriteRule ^a.js piwik.js [L,QSA]
RewriteRule ^a.php piwik.php [L,QSA]
```
and then when it provides you the code snippet should automatically have this url end point used.

**Proposed Solution**
$_GET variables can be json encoded and then encrypted with a key generated for each matomo install. You provide the key to the js tracking file. JavaScript then instead provides the data like so:
`a.php?VGhpcyBpcyBhbiBlbmNvZGVkIHN0cmluZw==`

On the PHP side you would simply base64 decode then json_decode then just array_merge with $_GET.

**Comment by arnowelzel** [comment]
The key needs to be embedded in the JavaScript - otherwise it is not possible to encryptthings on the client. So you don't gain any privacy, as anyone else could also send a request to the proxy script and find out, what key is used.

**Comment by select** [comment]
I managed to unblock uBlock by creating the symlinks for `piwik.(js|php) and changing pwik.js
```
"action_name=
```
to 
```
"a=b&action_name=
```
no need to rename variables @arnowelzel @bdore @londonuk371

**Comment by select** [comment]
What also might work is creating a symlink like 
```
mkdir foo
ln -s piwik.php foo/index.php
```
then point the tracking script to `/foo`, no need to mess with routes

**Comment by isopetalous** [comment]
Do not use piwik as the folder name, rename it to something else.
No need to rename the files or make rewrite rules.
Just modify the tracking code and replace all entries of "piwik.php" and "piwik.js" with "js/".
js/index.php will serve both php and js files for you.

**Comment by BurninLeo** [comment]
> If you use the "proxy" script to hide the real URL of your Piwik server (see https://github.com/piwik/tracker-proxy) you can also use that to hide parameter names as well.

That's just a cool idea, thanks! It's even possible without creating a new URL, if you just use `require('piwik.php')` below setting the `$_GET` variable.

For those who are still searching, here are two PHP scripts to put in the piwik directory:

pi-js.php
```
<?php
$jsFile = file_get_contents('piwik.js');
echo strtr($jsFile, [
  '"action_name="' => '"myAC="',
  'piwik.php' => 'pi-php.php'  // Not sure, if this is really necessary
]);
?>
```

pi-php.php
```
<?php
if (isset($_GET['myAC'])) {
        $_GET['action_name'] = $_GET['myAC'];
        unset($_GET['myAC']);
}
require('piwik.php');
?>
```

Then just replace `piwik.php` in the Matomo code by `pi-php.php` and `piwik.js` by `pi-js.php`. This solution should even survive the next update of Matomo.

**Comment by ston3o** [comment]
This works for me :)

```bash
mv matomo.js m.js && mv matomo.php m.php
mv piwik.js p.js && mv piwik.php p.php
sed -i -e 's/action_name=/a=b\&action_name=/g' m.js
```

```javascript
<script type="text/javascript">
  var _paq = window._paq || [];
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="//stats.domain.com/";
    _paq.push(['setTrackerUrl', u+'m.php']);
    _paq.push(['setSiteId', '1']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'m.js'; s.parentNode.insertBefore(g,s);
  })();
</script>
```

**Comment by w-biggs** [comment]
I found that the easiest thing to do overall is make two tiny changes:

1. Run `sed -i -e 's/action_name=/a=b\&action_name=/g' matomo.js` in the directory Matomo is installed in.
2. Replace the `g.src=u+'m.js'` in the tracking code with  `g.src=u+'js/'`.

That's it - now Matomo works, even if uBlock is installed.

**Comment by timbowhite** [comment]
@ston3o @w-biggs 

3. ```chmod 444 matomo.js``` or ```chattr +i matomo.js```

The [Auto-Archiving cron](https://matomo.org/docs/setup-auto-archiving/) runs periodically (either from your browser or via crontab depending on your config) and regenerates matomo.js, wiping out any changes you've made to it. The above should lock it down.

**Comment by mattab** [comment]
> regenerates matomo.js, wiping out any changes you've made to it. The above should lock it down.

To have your changes included in the regenerated matomo.js file, you can put the changes in the file `matomo/js/piwik.min.js` which is the "Source" file read to regenerate the minified file.

**Comment by aleksandar-todorovic** [comment]
There's an additional rule in EasyPrivacy:

```js
&idsite=*&send_image=$image
```

This makes it slightly trickier because doing a similar `sed` operation to the one mentioned by @ston3o and @w-biggs will not work thanks to that `*`. 

I guess the alternative approach would be to either rename one of those two variables or to switch the order of `send_image` and `idsite`.

**Comment by 45236** [comment]
> 
> 
> There's an additional rule in EasyPrivacy:
> 
> ```js
> &idsite=*&send_image=$image
> ```
> 
> This makes it slightly trickier because doing a similar `sed` operation to the one mentioned by @ston3o and @w-biggs will not work thanks to that `*`.
> 
> I guess the alternative approach would be to either rename one of those two variables or to switch the order of `send_image` and `idsite`.

that is exactly my problem. How do I rename that or how do I change the order?

**Comment by sgiehl** [comment]
@45236 you can use POST requests for tracking. That will make it impossible to replay tracking logs if something goes wrong with it, but should circumvent the rule. See https://developer.matomo.org/api-reference/tracking-javascript#advanced-uses

**Comment by arnowelzel** [comment]
JFTR: the modified proxy script still works. If you need to replace additional parameters, like "idsite" etc. - just go ahead and do it: https://arnowelzel.de/samples/piwik-tracker-proxy.txt

Live in action: https://arnowelzel.de

**Comment by ikmolbo** [comment]
> > There's an additional rule in EasyPrivacy:
> > ```js
> > &idsite=*&send_image=$image
> > ```
> > 
> > 
> > This makes it slightly trickier because doing a similar `sed` operation to the one mentioned by @ston3o and @w-biggs will not work thanks to that `*`.
> > I guess the alternative approach would be to either rename one of those two variables or to switch the order of `send_image` and `idsite`.
> 
> that is exactly my problem. How do I rename that or how do I change the order?

I managed to do this by chaining two search and replace commands as follows:

`sed -i -e 's/&send_image=0/\&/g' js/piwik.min.js; sed -i -e 's/action_name=/send_image=0\&action_name=/g' js/piwik.min.js`

This will reorder send_image=0 and also overcomes the filter on action_name. You will need to call `./console custom-matomo-js:update` once you have done this to ensure these changes are picked up.

Having done this, I am now seeing visits recorded even when using ad blockers.

**Comment by 45236** [comment]
@ikmolbo nice and in which file do I have to replace that?

**Comment by jb-ks** [comment]
@arnowelzel 
The proxy with your changes works like a charm, but I wonder how to get the opt-out working.

I know, that in your proxy-file (based on the old matomo-proxy) is no support für the opt-out frame.

So i tried to display the opt-out frame exactly as before using the proxy (accessing index.php in the matomo installation). The frame is displayed, you can opt-out and later opt-in again, but - even if it seems to be opted out - the tracking is running.

**Comment by arnowelzel** [comment]
@jb-ks Opt-out can't work because the ignore-cookie is set for the Matomo URL and not for the website which hosts the proxy script. The proxy script has no chance at all to check wether there is a ignore-cookie set or not, because the idea of this script is not to call the Matomo URL ever using the browser to avoid getting blocked.

In my case I just don't offer opt-out since I use Matomo without any cookies at all on my website:

```
try {
	var piwikTracker = Piwik.getTracker(pkBaseURL + "pwproxy.php", 1);
	piwikTracker.trackPageView();
	piwikTracker.enableLinkTracking();
	piwikTracker.disableCookies();
	} catch( err ) {}
```

Without any tracking cookies Matomo won't recognize returning visitors - but this is not important for me anyway. It's enough to see how many visits and page views I got and from which external URL they come (which only requires the HTTP referrer but no tracking cookies) and link tracking inside the website still works as this only requires the IP address during a session (all requests from the same IP address within a certain time period are considered to be the same visitor).

And yes - of course IP addresses are anonymized in Matomo on my server (the last two bytes are set to 0). But even anonymized IP addresses are sufficient for link tracking without tracking cookies.

And from a privacy standpoint: yes, I thinks this is legitimate, since accessing a website requires an active connection from the visitor anyway. A visitor can't say "I want to see your website but I don't want the webserver to know about that".

**Comment by jb-ks** [comment]
@arnowelzel 
Thanks for your explanation.

I have therefore decided to include a local alternative. The solution by Clive Beckett at [https://github.com/clivebeckett/matomo-opt-out](https://github.com/clivebeckett/matomo-opt-out) works very well after some initial detail problems.

**Comment by suppadeliux** [comment]
Hello, I tried all of the solutions, and somehow now my ad blocker is not blocking the request anymore!!

I created links with the command `ln -s piwik.js p.js && ln -s piwik.php p.php` & `ln -s matomo.js m.js && ln -s matomo.php m.php`, and also changed the action_name and idsite for something else.

And since I am using the angular plugin for matomo, I am overloading some of the code so I can change the `_paq.push(['setTrackerUrl', u + 'p.php']);` and the `g.src = !!scriptUrl ? scriptUrl : u + 'p.js';`

BUT, now I have this error:
`https://mysite.com/p.php?my_action.......... 403 (Forbidden)`

I guess that the tracker cannot find the file p.php. But i have no idea where to change this.

For example if I type in my browser the matomo instance `mysite.com/piwik.php` or `mysite.com/matomo.php` .
I have the message (This resource is part of Matomo. Keep full control of your data with the leading free and open source web analytics & conversion optimisation platform.)

But if I type `mysite.com/p.php`, this doesnt work and sends an error.

I am so confused that I dont really know what to do, maybe the solution is pretty easy but I cant see that.

Thank you for your help.

**Comment by Tealk** [comment]
I have renamed the files and also the action_name but now I get the following error and I don't understand why it is blocked.
![image](https://user-images.githubusercontent.com/12276250/106870378-55e1ee00-66d1-11eb-9162-89230f4c7117.png)

**Comment by Tealk** [comment]
I really do not know how to get around this
![grafik](https://user-images.githubusercontent.com/12276250/107117179-b2cbd880-6878-11eb-90dc-79b39a132558.png)

**Comment by arnowelzel** [comment]
> 
> 
> I really do not know how to get around this
> ![grafik](https://user-images.githubusercontent.com/12276250/107117179-b2cbd880-6878-11eb-90dc-79b39a132558.png)

By using a proxy script which takes parameters with other names and forwards this to the real script of Matomo as suggested already in 2017.

**Comment by Tealk** [comment]
> 
> By using a proxy script which takes parameters with other names and forwards this to the real script of Matomo as suggested already in 2017.

Does this still work, this is already a good 4 years old

**Comment by arnowelzel** [comment]
> 
> 
> > By using a proxy script which takes parameters with other names and forwards this to the real script of Matomo as suggested already in 2017.
> 
> Does this still work, this is already a good 4 years old

Yes - it still works, even with a current version of Matomo. The script I posted here is still in use on my own website and many others:

https://arnowelzel.de/samples/piwik-tracker-proxy.txt

**Comment by kylekatarnls** [comment]
It was even mentioned yet in 2015 ^^
https://github.com/matomo-org/matomo/issues/7364#issuecomment-140331043

Principle is that there are plenty different ad blockers trying to use different detection ways, you have to let nothing making possible to link the script to Matomo.

**Comment by Tealk** [comment]
> https://arnowelzel.de/samples/piwik-tracker-proxy.txt

but the tracking code seems to have changed and i am not really fit in javascript

```
<!-- Matomo -->
<script type="text/javascript">
    var _paq = window._paq = window._paq || [];
    _paq.push(['requireCookieConsent']);
    _paq.push(['setSecureCookie', true]);
    _paq.push(['trackAllContentImpressions']);
    (function() {
        var u = "https://trackedsite.com/";
        _paq.push(['setTrackerUrl', u + 'indexx.php']);
        _paq.push(['setSiteId', '6']);
        var d = document,
            g = d.createElement('script'),
            s = d.getElementsByTagName('script')[0];
        g.type = 'text/javascript';
        g.async = true;
        g.src = u + 'indexx.js';
        s.parentNode.insertBefore(g, s);
    })();
</script>
<noscript>
    <p><img src="https://trackedsite.com/indexx.php?idsite=6&amp;rec=1" style="border:0;" alt="" /></p>
</noscript>
<!-- End Matomo Code -->
```
how do I have to adjust this now?

**Comment by arnowelzel** [comment]
> > https://arnowelzel.de/samples/piwik-tracker-proxy.txt
> 
> but the tracking code seems to have changed and i am not really fit in javascript

The proxy script does not need to be changed - it just loads the content of `piwik.js` from Matomo and replaces a number of parameter names so they can be used with a different name.

On my site, the script is then used this way:

```
<script>
    var pkBaseURL = "https://arnowelzel.de/";
    document.write(unescape("%3Cscript src='" + pkBaseURL + "pwproxy.php' type='text/javascript'%3E%3C/script%3E"));
</script><script>
    try {
        var piwikTracker = Piwik.getTracker(pkBaseURL + "pwproxy.php", 1);
        piwikTracker.trackPageView();
        piwikTracker.enableLinkTracking();
        piwikTracker.disableCookies();
    } catch( err ) {}
</script><noscript><p><img src="https://arnowelzel.de/pwproxy.php?ids=1&amp;rec=1" style="border:0" alt="" /></p></noscript>
```

And this works with Matomo just fine. No need for any JavaScript changes. Just put `pwproxy.php` with the content as shown in https://arnowelzel.de/samples/piwik-tracker-proxy.txt and use it as shown above.

I also have page which shows the statistics of my site by using the Matomo API: https://arnowelzel.de/en/tools/statistics

**Comment by Tealk** [comment]
How do I include `_paq.push(['requireCookieConsent']);` there then?

**Comment by Tealk** [comment]
is it possible to enable `_paq.push(['setSecureCookie', true]);`?

and i have some functions like:
```
function showHome() {
  ...
  _paq.push(['setDocumentTitle', 'Home']);
  _paq.push(['trackPageView']);
}
```
Also I get the message and I do not want to turn off nosniff
`The resource of "https://trackedsite.com/mproxy.php" was blocked because of a MIME type conflict ("text/html") (X-Content-Type-Options: nosniff).`

**Comment by arnowelzel** [comment]
You can substitude the `_paq` calls with the respective Methods of `piwikTracker`, e.g.

```
piwikTracker.setDocumentTitle('Home');
piwikTracker.trackPageView();
``` 

Note: `piwikTracker.trackPageView();` is already included in my code.

See the available methods of the "old" API at https://developer.matomo.org/api-reference/tracking-javascript.

About the cookie consent: I don't set a cookie at all (`piwikTracker.disableCookies();`), so I also don't need any consent for this. In my experience this is good enough to get useful statistics for a website. I don't need information about returning visitors. However - if you use cookies you should also implement some kind of cookie consent dialog on your website to fulfil GDPR requirements.

**Comment by Tealk** [comment]
> However - if you use cookies you should also implement some kind of cookie consent dialog on your website to fulfil GDPR requirements.

I have written a very nice one for myself.

**Comment by omarr1000** [comment]
@arnowelzel 
I just added your script (piwik-tracker-proxy.txt) and the js on the site. So far fine, no errors etc. 
BUT in my Matomo nothing is tracking! Any idea?

```
var pkBaseURL = "https://sayami.de/";
document.write(unescape("%3Cscript src='" + pkBaseURL + "pwproxy.php' type='text/javascript'%3E%3C/script%3E"));
</script><script>
try {
  var piwikTracker = Piwik.getTracker(pkBaseURL + "pwproxy.php", 1);
  piwikTracker.trackPageView();
  piwikTracker.enableLinkTracking();
  piwikTracker.disableCookies();
} catch( err ) {}
```

![grafik](https://user-images.githubusercontent.com/32812277/135625579-2e91d333-36a3-42b3-bb69-731d4d7cb716.png)


The default tracking code did work...

thx in advance (-;

**Comment by arnowelzel** [comment]
@omarr1000 The script has to be named `pwproxy.php`, so you can call it using `https://sayami.de/pwroxy.php`. Also you need to need to adjust the script so it uses the correct tracking IDs from your server.

**Comment by omarr1000** [comment]
sure I know, all done as you mentioned, is named pwproxy.php, in that file edited the correct path and auth_token 

$PIWIK_URL = 'https://sayami.de/path_to_matomo/';
$TOKEN_AUTH = 'xxxyyyyzzzz';

with tracking ID you mean the SiteId? or the token_auth?

**Comment by arnowelzel** [comment]
_(forget the comment about pwproxy.php - seems to be ok now)_

And yes, I talk about the TOKEN_AUTH, just follow the instructions in the script.

And is  https://sayami.de/path_to_matomo/ the correct path? If you run Matomo on another domain you have to put the domain here not some path inside sayami.de

**Comment by omarr1000** [comment]
The Matomo path is in the same domain as described

**Comment by arnowelzel** [comment]
Then I don't know what's wrong, sorry.

**Comment by jsvtrlvrslv** [comment]
The 

> > > https://arnowelzel.de/samples/piwik-tracker-proxy.txt
> > 
> > 
> > but the tracking code seems to have changed and i am not really fit in javascript
> 
> The proxy script does not need to be changed - it just loads the content of `piwik.js` from Matomo and replaces a number of parameter names so they can be used with a different name.
> 
> On my site, the script is then used this way:
> 
> ```
> <script>
>     var pkBaseURL = "https://arnowelzel.de/";
>     document.write(unescape("%3Cscript src='" + pkBaseURL + "pwproxy.php' type='text/javascript'%3E%3C/script%3E"));
> </script><script>
>     try {
>         var piwikTracker = Piwik.getTracker(pkBaseURL + "pwproxy.php", 1);
>         piwikTracker.trackPageView();
>         piwikTracker.enableLinkTracking();
>         piwikTracker.disableCookies();
>     } catch( err ) {}
> </script><noscript><p><img src="https://arnowelzel.de/pwproxy.php?ids=1&amp;rec=1" style="border:0" alt="" /></p></noscript>
> ```
> 
> And this works with Matomo just fine. No need for any JavaScript changes. Just put `pwproxy.php` with the content as shown in https://arnowelzel.de/samples/piwik-tracker-proxy.txt and use it as shown above.
> 
> I also have page which shows the statistics of my site by using the Matomo API: https://arnowelzel.de/en/tools/statistics

Does this trick still work in the latest version of Matomo?

**Comment by arnowelzel** [comment]
Hi,

the tracking protection in some browsers may detect the JavaScript calls 
of Matomo even if URLs and variable names of obfuscated because of the 
way how the scripts of Matomo behave. For example Vivaldi will block 
requests by the Matomo script as malicous "ping" request.

So, yes, the script still works in the way that it can be used with the 
latest version of Matomo as well (it is still in use this way on my own 
website), but Matomo may get blocked anyway.

All the best,
Arno

On 11.01.2023 00:17, jsvtrlvrslv wrote:

> The
> 
> https://arnowelzel.de/samples/piwik-tracker-proxy.txt
> 
> but the tracking code seems to have changed and i am not really fit in 
> javascript

The proxy script does not need to be changed - it just loads the content 
of piwik.js from Matomo and replaces a number of parameter names so they 
can be used with a different name.

On my site, the script is then used this way:

<script>
     var pkBaseURL = "https://arnowelzel.de/";
     document.write(unescape("%3Cscript src='" + pkBaseURL + 
"pwproxy.php' type='text/javascript'%3E%3C/script%3E"));
</script><script>
     try {
         var piwikTracker = Piwik.getTracker(pkBaseURL + "pwproxy.php", 
1);
         piwikTracker.trackPageView();
         piwikTracker.enableLinkTracking();
         piwikTracker.disableCookies();
     } catch( err ) {}
</script><noscript><p><img 
src="https://arnowelzel.de/pwproxy.php?ids=1&amp;rec=1" style="border:0" 
alt="" /></p></noscript>

And this works with Matomo just fine. No need for any JavaScript 
changes. Just put pwproxy.php with the content as shown in 
https://arnowelzel.de/samples/piwik-tracker-proxy.txt and use it as 
shown above.

I also have page which shows the statistics of my site by using the 
Matomo API: https://arnowelzel.de/en/tools/statistics

Does this trick still work in the latest version of Matomo?

--
Reply to this email directly, view it on GitHub [1], or unsubscribe [2].
You are receiving this because you were mentioned.Message ID: 
***@***.***>



Links:
------
[1] 
https://github.com/matomo-org/matomo/issues/7364#issuecomment-1378021626
[2]

**Comment by elmaxid** [comment]
Hello,

I'm playing with Matomo on premise and I found that some ad blockers like Adblock, ublock, etc. are blocking Matomo's JavaScript.

Let me tell you how I solved it and apparently it's been working fine for a few days.

1) I have Matomo installed on a server with nginx.

The hack consists of changing the URL of the requests, for which I did the following:

**Within the root directory of Matomo:**

2) Created a directory called "_mt_" within the root directory:

`mkdir mt`

3) Created symbolic links of the _js_ and _php_:

```
ln -s matomo.js mat.js
ln -s matomo.php mt/index.php
```

Make sure that the last "_matomo.php_" has a symbolic link created inside the "_mt_" directory.

Added to the **nginx** configuration that the "_mt_" directory serves the index.php directly.
```

location /mt {
   index index.php;
 }
```
4) Modify the script like:

```

<script>
          var _paq = window._paq = window._paq || [];      
          _paq.push(['trackPageView']);
          _paq.push(['enableLinkTracking']);
          (function() {
            var u="//xxxx.xxx.com/";
            _paq.push(['setTrackerUrl', u+'mt/']);
            _paq.push(['setSiteId', '3588']);
            var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
            g.async=true; g.src=u+'mat.js'; s.parentNode.insertBefore(g,s);
          })();
        </script>
```


and now works with adblock and ublock without problem

I hope this help

M.

**Comment by foxx9** [comment]
Hi, 
The simplest way is to use what @Joey3000 mentioned : 
https://github.com/piwik/piwik/blob/master/js/README.md
 
Use `js/` both as the javascript  and php url  **AND** enable POST requests to avoid parameters filtering:

``` 
...
_paq.push(['setTrackerUrl', u+'js/']);
....
_paq.push(['setRequestMethod', 'POST']);
...
g.async=true; g.src=u+'js/'; s.parentNode.insertBefore(g,s);
...
```

**Comment by twarkie** [comment]
> Hi, The simplest way is to use what @Joey3000 mentioned : https://github.com/piwik/piwik/blob/master/js/README.md
> 
> Use `js/` both as the javascript and php url **AND** enable POST requests to avoid parameters filtering:
> 
> ```
> ...
> _paq.push(['setTrackerUrl', u+'js/']);
> ....
> _paq.push(['setRequestMethod', 'POST']);
> ...
> g.async=true; g.src=u+'js/'; s.parentNode.insertBefore(g,s);
> ...
> ```

This almost works. The script is loaded just fine even with adblock, but the actual tracking request is still blocked. Not sure why but maybe the url parameters are filtered somehow. Having `_paq.push(['setRequestMethod', 'POST']);` seems to keep all parameters in the URL even though the request is sent using `POST`. Would be nice to have a clean URL and move the url params to post data.

**Comment by arnowelzel** [comment]
@twarkie Modern blockers do not only work based on URL patterns but also recognize the behaviour of trackers like Piwik/Matomo. For example uBlock Origin recognizes that as "tracker ping" request and will block that, no matter how you name parameters or URL. This is something you can not overcome.

**Comment by Fransferdy** [comment]
It is 2024, I'm trying to use Matomo, I use the browser Brave here. Well, Brave sure seems to be awesome at privacy, because it blocks matomo no matter what I do!
1. I changed directories (matomo) to analise.
2. I changed matomo.js to analise.js, and matomo.php to batatinha.php
3. Changed action_name to master_potato
At this point, Brave still blocks the php call, but not the js files.

4. I went into the js sendRequest function, and replaced the query parameters with a Reverse, base64Encoded version of themselves, something like query = base64Encode(query).reverse()
5. In the anali.sephp(matomo.php), I unreverse and decode the base64, before setting the query parameters for PHP.

The above is working correctly, but somehow Brave still manages to find out it is a tracking request and blocks it.

So all files have different names, the query parameters are completly different and obfuscated, and still, Brave manages to keep my privacy(and everyone elses! haha). I don't know what else I can do.

Also, the default matomo.js script seems to only count Page Loads, so it does not really work with React Websites, Matomo does have an alternative to that, which is the Tag Manager, that generates and regenerates a custom matomo.js, 'destroying' all the possible corrections we have been doing in this thread, any ideas ?
At this point I think that if you want to avoid adblockes and use analytics with a Reactive Website, maybe Matono isn't ideal.

**Comment by arnowelzel** [comment]
Modern blockers stop Matomo and other access tracking tools because of their behaviour and not because of specific URLs - and as far I know there is no way to circumvent this.

**Comment by jsvtrlvrslv** [comment]
> Modern blockers stop Matomo and other access tracking tools because of their behaviour and not because of specific URLs - and as far I know there is no way to circumvent this.

Do you have any material that proves this? 

If you use a script on the same domain, pass the .php and only then, via the server side, send the .php to Matomo (external domain)... will this not solve the problem?

**Comment by arnowelzel** [comment]
Just look at the network requests list in the web developer tools and look for type "ping" requests which get blocked, for example with uBlock Origin enabled. This kind of block can not be circumvented since the behaviour is blocked, not an URL.

**Comment by Fransferdy** [comment]
> > Modern blockers stop Matomo and other access tracking tools because of their behaviour and not because of specific URLs - and as far I know there is no way to circumvent this.
> 
> Do you have any material that proves this?
> 
> If you use a script on the same domain, pass the .php and only then, via the server side, send the .php to Matomo (external domain)... will this not solve the problem?

That's the next thing I will try, I just need to find out how to get the stats from matomo.js so I can pack it and send it to my server in each page load and history change. 

I wonder if Brave marks variables with tracking/user information as "tainted" somehow and any attempt to "send" them out of the browser is blocked.(it would be really cool if they did that, and hard to manage lol)

**Comment by jsvtrlvrslv** [comment]
Também vou tentar isso. Quando tiveres novidades partilha!

**Comment by Teisi** [comment]
I believe that bypassing uBlock, for example, would violate the GDPR, as the user obviously explicitly does not want to be tracked. This would therefore ultimately be treated as an opt-out. (The GDPR requires an explicit opt-in).... so.... (just an opinion)

**Comment by arnowelzel** [comment]
Just counting how many people accessed certain URLs on a website is not "tracking".

Tracking starts, when you also set a cookie to recognize the users when the come back again later. But when you disable the cookies of Matomo (which is possible) and anonymize the IP addresses, this should not be an issue for GDPR.

I always use Matomo this way: no tracking cookies, only anonymized IP addresses on the Matomo server:

```
<script nonce="... nonce for CSP policy ...">
    var _paq = window._paq = window._paq || [];
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    _paq.push(['disableCookies']);
(function() {
        var u="https://my-matomo-server.example/";
        _paq.push(['setTrackerUrl', u+'matomo.php']);
        _paq.push(['setSiteId', '1']);
        var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
        g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
    })();
</script>
```

If needed, you can also use `_paq.push(['enableHeartBeatTimer', 5]);` to enable a "heartbeat timer" with 5 second interval (the interval can also be higher, you may have to test different values) to avoid Matomo detecting many visitors as leaving your site within a few seconds, because they don't send a cookie with the next request any longer.

**Comment by Teisi** [comment]
@arnowelzel technically the process of making the IP address anonym isn't allowed by default only after opt-in - i know nobody cares about that...
It only says something about processing and nothing about "saving", and also
transparency, which means I have to point this out to the user somehow.):
1 (a):
https://gdpr-info.eu/art-5-gdpr/
or
https://commission.europa.eu/law/law-topic/data-protection/rules-business-and-organisations/principles-gdpr/overview-principles/what-data-can-we-process-and-under-which-conditions_en

But it actually goes even further: “the company/organization must collect and process only the personal data that is necessary to fulfil that purpose (‘data minimization’);” The purpose without the user's consent is to view and use the website, and not to be “tracked”.


"Just counting how many people accessed certain URLs on a website is not "tracking"."
Unfortunately, I have to disagree, due to the fact that you need personalized information (e.g. IP address and or browser information), which is already considered “personalized” in itself, to find out how many people have accessed the server.
If you only want to track access without knowing how many people have accessed the server, then this is possible - but you don't need Matomo for this. :) . (with the technical note above in the mind)

In a way there is a “gray” zone -> server-logs . Server logs themselves are allowed due to other legal requirements (protection of the server: spam, hackers, etc.), but should not actually be merged with other data (e.g. from Matomo).
And I would question whether the evaluation of these logs for any other purpose than protection is permitted.


-----------------
Off-topic:

ohhh if only I had the money.... e-mail traffic is a big data protection hole... but nobody cares about GDPR anyway, so... what do i know. xD muharhar

**Comment by arnowelzel** [comment]
@Teisi 

> [@arnowelzel](https://github.com/arnowelzel) technically the process of making the IP address anonym isn't allowed by default only after opt-in - i know nobody cares about that... It only says something about processing and nothing about "saving", and also transparency, which means I have to point this out to the user somehow.): 1 (a): https://gdpr-info.eu/art-5-gdpr/ or https://commission.europa.eu/law/law-topic/data-protection/rules-business-and-organisations/principles-gdpr/overview-principles/what-data-can-we-process-and-under-which-conditions_en

I think, you misunderstood me.

By default you are **not allowed** to store any personal data at all. So the opposite is true: by default you **must not** store full IP addresses without user consent. But anonymizing the data, so the user can **not** be identified in any way, is of course allowed.

When I say "anonymize" I mean modifying the data in a way that it it is absolutely impossible to track a user based on this. If neccessary this means **removing** the address completely (which is also possible in Matomo using plugins). For the long term statistics there is no need to keep any individual log record at all and Matomo will delete those records anyway after processing the long term statistics to save data.

> But it actually goes even further: “the company/organization must collect and process only the personal data that is necessary to fulfil that purpose (‘data minimization’);” The purpose without the user's consent is to view and use the website, and not to be “tracked”.

Exactly! And "tracking" does not happen if you do **not** store full IP adresses or no IP adresses at all.

> "Just counting how many people accessed certain URLs on a website is not "tracking"." Unfortunately, I have to disagree, due to the fact that you need personalized information (e.g. IP address and or browser information), which is already considered 
“personalized” in itself, to find out how many people have accessed the server. If you only want to track access without knowing how many people have accessed the server, then this is possible - but you don't need Matomo for this. :) . (with the technical note above in the mind)

No ,you do **not** need IP addresses to do counting. You can even **remove** every information except the request URI completely. Then you just have a list of URIs and you are still able to do create a statistic. You don't even need timestamps - just delete the list every day at 0:00 and by the end of the day do the counting before you delete the list again. Then you have a number of requests for each URI every day without storing any data about users at all.

Is a counter on a page like "this site was visited nnn times" where nnn will automatically go up by 1 for every HTTP GET request already "tracking"? I don't think so.

> In a way there is a “gray” zone -> server-logs . Server logs themselves are allowed due to other legal requirements (protection of the server: spam, hackers, etc.), but should not actually be merged with other data (e.g. from Matomo). And I would question whether the evaluation of these logs for any other purpose than protection is permitted.

As I said above: you can even log without IP address. And you can also make sure, that the IP address or the complete log is removed after a certain time period, for example after 30 or 60 days.

And Art. 5 GDPR only applies if you *do* store personal data. A list of URIs without any other information is not "personal data" since there is nothing you can use to find out which person requested that URI (of course the URI itself must not contain personal data then). Also anonymized IP addresses is *not* personal data. If course Art. 5 GDPR does *not* force you to log every access to your webserver and first ask people if they would agree to protect their privacy by removing the full IP address from the log.

**Comment by arnowelzel** [comment]
@Teisi 

Quote:

> the company/organisation must collect and process only the personal data that is necessary to fulfil that purpose (‘data minimisation’);
> the company/organisation must ensure the personal data is accurate and up-to-date, having regard to the purposes for which it is processed, and correct it if not (‘accuracy’);

Solution:

Do not store personal data at all, then the second requirement is already obsolete. There is no data.

**Comment by Teisi** [comment]
yes i think we are talking past each other :)

Even changing the IP address to a shortened version, for example, is already a “process” - at this point you haven't saved anything yet, but according to the GDPR text this is actually not allowed.
(“IP address” is only a placeholder for all other data that can be assigned to a user - by the way, this also applies to data that is not personal on its own but is personal again when combined with other (non-personal) data - but that's another topic :D )

And as I said, if you want to know how many USERS (not just hits) have visited the website, you have to process user data in some way. Even if you don't save it.
That's the whole crux of the legal text :D


To clarify:
Changing string, array, integer, and so on is already processing.

And of course: no plaintiff, no judge, no executioner ;)

Therefore only my opinion, taking into account that I am also a “user”. XD

**Comment by arnowelzel** [comment]
@Teisi 

> yes i think we are talking past each other :)
> 
> Even changing the IP address to a shortened version, for example, is already a “process” - at this point you haven't saved anything yet, but according to the GDPR text this is actually not allowed. (“IP address” is only a placeholder for all other data that can be assigned to a user - by the way, this also applies to data that is not personal on its own but is personal again when combined with other (non-personal) data - but that's another topic :D )

There is no IP address when you don't store it.

```
<?php
$url = $_SERVER['REQUEST_URI'];
$timestamp = date('Y-m-d H:i:s');

// now log $url and $timestamp ...
```

You really want to claim, that GDPR requires you to **store** the IP full address until any visitor allows you to not do so?

> To clarify: Changing string, array, integer, and so on is already processing.

Which is irrelevant, when that data is **not** stored. GDPR is **only** relevant when you **store** data. As soon as IP addresses are not stored at all or only in an anonymized version, so that they are **not** personal data any longer, GDPR does not apply any longer. Otherwise the existence of NAT without consent of the users would be illegal since NAT clearly changes IP addresses without asking the user first.

**Comment by Teisi** [comment]
> > To clarify: Changing string, array, integer, and so on is already processing.
> 
> Which is irrelevant, when that data is **not** stored. GDPR is **only** relevant when you **store** data. As soon as IP addresses are not stored at all or only in an anonymized version, so that they are **not** personal data any longer, GDPR does not apply any longer. Otherwise the existence of NAT without consent of the users would be illegal since NAT clearly changes IP addresses without asking the user first.

And that's where our opinions diverge :)
I am of the opinion that it does not depend on whether I store any personal data or not, but that the GDPR already prohibits the processing alone. In the GDPR text (from above) it is explicitly written only “processing” and not “storing”.

German:
Auch im Deutschen steht nur "verarbeiten" / "weiterverarbeiten" - von Speichern steht auch dort nichts:
https://dsgvo-gesetz.de/art-5-dsgvo/

**Comment by arnowelzel** [comment]
> > > To clarify: Changing string, array, integer, and so on is already processing.
> > 
> > 
> > Which is irrelevant, when that data is **not** stored. GDPR is **only** relevant when you **store** data. As soon as IP addresses are not stored at all or only in an anonymized version, so that they are **not** personal data any longer, GDPR does not apply any longer. Otherwise the existence of NAT without consent of the users would be illegal since NAT clearly changes IP addresses without asking the user first.
> 
> And that's where our opinions diverge :) I am of the opinion that it does not depend on whether I store any personal data or not, but that the GDPR already prohibits the processing alone. In the GDPR text (from above) it is explicitly written only “processing” and not “storing”.
> 
> German: Auch im Deutschen steht nur "verarbeiten" / "weiterverarbeiten" - von Speichern steht auch dort nichts: https://dsgvo-gesetz.de/art-5-dsgvo/

See my example from above - is this processing or not?

```
<?php
$url = $_SERVER['REQUEST_URI'];
$timestamp = date('Y-m-d H:i:s');

// now log $url and $timestamp ...
```

Do I have to add the IP address or not? Because according to your opinion it is *forbidden* to leave the IP address out, if PHP knows it in `$_SERVER['REMOTE_ADDR']`. There is no difference between modifying the address to not recording it at all. Because "not recording it at all" is identical to "removing parts of it". In any case, the IP address was "processed" since PHP itself already knows the address.

Edit: also see here - https://gdpr-info.eu/art-4-gdpr/

> ‘processing’ means any operation or set of operations which is performed on personal data or on sets of personal data, whether or not by automated means, such as collection, recording, organisation, structuring, storage, adaptation or alteration, retrieval, consultation, use, disclosure by transmission, dissemination or otherwise making available, alignment or combination, restriction, erasure or destruction;

So "storage" is also part of "processing". And of course the whole GDPR is about storing data. If data is not stored, there is nothing to discuss about.

**Comment by Teisi** [comment]
I would still want to speak again, at least in part:

(a) the IP is probably technically necessary to fulfill the task (the user's request) - I'm not that deep into the matter now, but theoretically only the web server itself would need to know the IP, maybe necessary for PHP sessions and php streams / sockets?!

(b) yes something like “$url = $_SERVER[‘REQUEST_URI’];” is already processing. Because you are doing something with the data, storing it in a variable (even if it is only stored for runtime).

>  And of course the whole GDPR is about storing data. If data is not stored, there is nothing to discuss about.

a) if you do not store any data, then you don't need matomo or similar software. But that would make the entire thread here unnecessary.

b) the GDPR applies to personal data and to data that can be retroactively assigned to a user. In other words, if you want to know how many users have visited your website, you need to know which request comes from which ONE user, so you need to store enough data (no matter which) that can be clearly assigned to this user (over multiple request / time). <- this requires consent.
If you do not do this, then you can only count requests per se, but these do not provide any information about how many users have visited your site.
So even if you only store a salted, hashed ID, as long as you can explicitly assign it to a user, this is personal and therefore a case for the GDPR.


This is where the other competitors come into play, such as "plausible.io", who are probably trying to circumvent the whole thing in this way:
https://plausible.io/data-policy#how-we-count-unique-users-without-cookies
Although this also has the disadvantages described there -> not possible across devices, "new" users every day, etc.

`hash(daily_salt + website_domain + ip_address + user_agent)`
But for a whole day, the user can be clearly identified. However, the GDPR does not prescribe a period of time in which this would be permitted, but prohibits it completely and utterly

So i say: I dare to doubt that this would stand up to a court ruling, as the IP address (at least in Germany is sometimes still changed daily) and the IP address is still considered personal.

**Comment by arnowelzel** [comment]
> I would still want to speak again, at least in part:
> 
> (a) the IP is probably technically necessary to fulfill the task (the user's request) - I'm not that deep into the matter now, but theoretically only the web server itself would need to know the IP, maybe necessary for PHP sessions and php streams / sockets?!
> 
> (b) yes something like “$url = $_SERVER[‘REQUEST_URI’];” is already processing. Because you are doing something with the data, storing it in a variable (even if it is only stored for runtime).

Yes, but `$_SERVER[‘REQUEST_URI’]` is not personal data.

> > And of course the whole GDPR is about storing data. If data is not stored, there is nothing to discuss about.
> 
> a) if you do not store any data, then you don't need matomo or similar software. But that would make the entire thread here unnecessary.

Exactly! That's what I talk about! If you just count how many times a URI was called - **without** recording **who** called it - then you don't have personal data!

[...]
> So i say: I dare to doubt that this would stand up to a court ruling, as the IP address (at least in Germany is sometimes still changed daily) and the IP address is still considered personal.

Why do you always bring up IP addresses? There is no need for that! See my example from above!

The log will look like this:

```
2025-07-20 10:23:00 https://domain.example/about
2025-07-20 10:23:02 https://domain.example/products
2025-07-20 10:23:07 https://domain.example/about
2025-07-20 10:24:10 https://domain.example/somethingelse
2025-07-20 10:26:00 https://domain.example/products
```

This will tell you this for 2025-07-20:

```
2x https://domain.example/about
2x https://domain.example/products
1x https://domain.example/somethingelse
```

So no personal data at all.

And yes, this is possible with Matomo as well.
